### ISSUE TITLE
regression: wrong result when joining with an unmaterialized source 

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#11072]
### What version of Materialize are you using?

v0.22.1-dev (9611cc695)

### How did you install Materialize?

Docker image

### What was the issue?

The `joins.td` test fails repeatably as follows:

```
> CREATE SOURCE mods_unmat FROM KAFKA BROKER 'kafka:9092' TOPIC 'testdrive-mods-3568222347' FORMAT AVRO USING SCHEMA '{   "type": "record",   "name": "envelope",   "fields": [     {       "name": "before",       "type": [         {           "name": "row",           "type": "record",           "fields": [             {"name": "num", "type": "long"},             {"name": "mod", "type": "string"}           ]         },         "null"       ]     },     { "name": "after", "type": ["row", "null"] }   ] }' ENVELOPE DEBEZIUM
--
  | rows match; continuing at ts 1646637643.033189
  | > CREATE VIEW test15 (names_num, names_name, mods_num, mods_mod) AS SELECT * FROM names FULL OUTER JOIN mods_unmat ON 1 = 0;
  | rows match; continuing at ts 1646637643.0903223
  | > SELECT * FROM test15;
  | rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 112ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 893ms
  | joins.td:237:1: error: non-matching rows: expected:
  | [["1", "one", "<null>", "<null>"], ["2", "two", "<null>", "<null>"], ["3", "three", "<null>", "<null>"], ["<null>", "<null>", "0", "even"], ["<null>", "<null>", "1", "odd"], ["<null>", "<null>", "2", "even"]]
  | got:
  | [["1", "one", "<null>", "<null>"], ["2", "two", "<null>", "<null>"], ["3", "three", "<null>", "<null>"]]
  | Diff:
  | row missing: ["<null>", "<null>", "0", "even"]
  | row missing: ["<null>", "<null>", "1", "odd"]
  | row missing: ["<null>", "<null>", "2", "even"]
  |  
  | \|
  | 236 \|
  | 237 \| > SELECT * FROM test15;
  | \| ^

```

The failure is:
- permanent, that is, the result never becomes correct no matter the timeout/retries
- goes away if the source is dropped and recreated

This is seen both on the `--workers 1` and the `cluster` Nightly tests. I am able to reproduce it by running the following loop on a `c5.2xlarge` machine:

```
 while bash -c -e "./mzcompose down -v ; ./mzcompose run default --workers 1 joins.td" ; do : ; done
```

Bisection reports the following commit:

```
3af43e25388609cb43b7ab352ffe03d34f89d552 is the first bad commit
commit 3af43e25388609cb43b7ab352ffe03d34f89d552
Author: Nikhil Benesch <nikhil.benesch@gmail.com>
Date:   Sat Mar 5 14:53:30 2022 -0500

    coord: allow SELECTing from unmaterialized sources
```

### Relevant log output

```source
No errors in the Mz logs.
```

**Comment by guswynn** [comment]
@philip-stoev How often did you see this fail? I am running in repeatedly (not using mzcompose, but with `--workers 1` and with a fresh mz and mzdata each time), and I haven't seen a failure with like 10+ runs

**Comment by guswynn** [comment]
not when testing locally, when running ` while bash -c -e "./mzcompose down -v ; ./mzcompose run default --workers 1 joins.td" ; do : ; done` on a `dev-box` (t3.2xlarge, not a c5.2xlarge, but I can't imagine that matters), I still dont see a repro. The nightly failed with this test this morning, on this commit: https://github.com/MaterializeInc/materialize/commit/2ab33038daf6643c07ffd2cf73c2d5ce327d3d9f
I dont see any relevant changes on main past that

**Comment by guswynn** [comment]
`release` build to trigger this WAY more, adding a sleep in various places seems to help as well, but in some cases it appears to be consistently triggered even without any sleeping, perhaps my kafka setup is just in the right place locally to trigger it

**Comment by guswynn** [comment]
This is the `EXPLAIN` I get:
Source materialize.public.mods_unmat (u11):
| Project (#0, #1)

Query:
%0 =
| Get materialize.public.names (u7)
| Map null, null

%1 =
| Get materialize.public.mods_unmat (u11)
| Map null, null
| Project (#2, #3, #0, #1)

%2 =
| Union %0 %1

to me this looks like (correctly, it noticed 0=1 is always false) it turned the RAW plan:
%0 =
| Get materialize.public.names (u13)

%1 =
| Get materialize.public.mods_unmat (u17)

%2 =
| FullOuterJoin %0 %1 on (1 = 0)
into just a union of 2 projections, where the first adds the null, null to the materialized source (I assume using its index), and the second…builds an index for the correct columns in the unmaterialized source, as that doesn't have an index yet? Does this all make sense? when I squint it seems right

**Comment by nmeagan11** [comment]
Is this still a `P1` bug? It's been hanging around for a bit now. If so, we should address ASAP.

**Comment by benesch** [comment]
The P1-ness was resolved by #11244. The remaining issue is that it is currently "in spec" for unmaterialized sources to return no data if you `SELECT unmat_src AS OF now()`. #11048 is tracking improving that, so I think we can call this one closed.
